Modify present pass managers for control-flow support

jakelishman opened the pull request:
### Summary

This adds control-flow support to the level 0 and level 1 pass managers directly, including neatly erroring out at runtime if the pass manager was constructed with options that are not yet compatible with control flow.  In general, we assume that plugins from outside of Terra are safe for control flow - it's highly likely that this is not in fact the case, but it's better to do things this way so an external plugin _can_ support control flow, as opposed to forbidding them from doing so.

Level 2 and level 3 both contain optimisations that are incompatible with control flow right now, and so immediately error if an input circuit has control flow.  The failing optimisations are ones that involve a two-pass process of "block collection" followed by "block replacement"; we have no updated any of the "block collection" analysis passes to be safe for control flow, since we would also likely need to communicate the control-flow structure through the property set, which is not a path we're going down just yet.

In order to construct pass managers with nested flow control (e.g. the routing-pass stages contain conditionals, but the entire routing stage needs to be conditioned on the presence of control flow), there is a new `PassManager` method `to_flow_controller`.  This produces a `FlowControllerLinear` with the same content as the pass manager, such that it can then itself be conditioned as a complete block, and embedded into other pass managers.

<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->


### Details and comments

Depends (effectively) on #8418 - the modifications to level 1 assume that `StochasticSwap` will work with control flow.

I haven't yet written tests for this, as I'm not 100% certain this is the right direction and I don't want to waste time.  Most of them likely wouldn't pass without #8418 anyway.  I'll write tests once we've got a bit of an agreement in spirit.

I anticipate this PR being the one that will add all the release notes for the new control-flow support through the transpiler - I will also add those before merge.

Fix #8630. (This will be the last PR of that epic.)

mtreinish commented:
So I haven't looked at this in too much detail but we did add support for nested #6962. I was able to use that to execute a series of passes with a nested controller to enable `GatesInBasis` to trigger basis translation in the optimization loop without needing `PassManger` changes. For example: https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/transpiler/preset_passmanagers/level2.py#L246-L249 (where `unroll` is built from a `PassManager` used for the translation stage here: https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/transpiler/preset_passmanagers/level2.py#L240 ). I won't say it's the most ergonomic interface, but in the interest of keeping the API changes minimal at this stage in the release could you use something like that to do what you're doing with `to_flow_controller`?

jakelishman commented:
The majority of the failing tests here transpired to be because of two reasons:
- the `is not None` test in evaluating whether the trivial layout was perfect (which also decides whether `VF2PostLayout` should run) was important when no layout stage had run.  Fixed by ff05b66.
- #6962, which added support for nested `FlowController`s didn't propagate the resultant `DAGCircuit` instances from its passes - they might "accidentally" work if the inner pass mutated its input rather than returning a new DAG, but this broke down for more complex passes.  Fixed by a363da7.

kdk commented on qiskit/transpiler/preset_passmanagers/level1.py:
```diff
@@ -94,15 +96,9 @@ def _choose_layout_condition(property_set):
-        return False
+        # Verify that a trivial layout is perfect.  If perfect, then the circuit should already
+        # appear to be swap mapped.
+        return property_set["is_swap_mapped"] is not None and not property_set["is_swap_mapped"]
```
Pre-existing, but the above comment still seems inconsistent. This seems to be checking both that `CheckMap` ran and that it didn't find the trivial layout to be perfect.

jakelishman replied:
Done in f1a0bbab2.

mtreinish commented on qiskit/transpiler/preset_passmanagers/level2.py:
```diff
@@ -262,12 +262,15 @@ def _unroll_condition(property_set):
         sched = plugin_manager.get_passmanager_stage(
             "scheduling", scheduling_method, pass_manager_config, optimization_level=2
         )
+    init = common.generate_error_on_control_flow(
+        "The optimizations in optimization_level=2 do not yet support control flow."
+    )
```
We technically could get a level2 pm that given the assumptions made in `common.py` is runnable with control flow. But it'd basically be `layout_method=dense`, `routing_method=stochastic`, `optimization=plugin`. It seems to be fairly unlikely in practice so this is fine I think.

jakelishman replied:
I was treating us as "supporting" an optimisation level if the defaults will work - for O2, you have to override everything that actually makes it O2 back to what's in O1, so I think it's better to reject `optimization_level=2` out-of-hand.

jakelishman replied:
I guess at O1, now that routing and layout are both Sabre by default there's an argument that we're kind of just doing that internally too, I suppose.  I was still thinking in terms of the old O1 that had dense/stochastic as its default.

mtreinish replied:
I'm fine with this for now, to a certain degree if people want to get more complicated maybe they should just be using custom pass managers. We just will have to document clearly in the release note that `optimization_level` 2 and 3 will error regardless of options with a control flow circuit.

kdk reviewed (commented):
Overall approach LGTM and matches what we discussed offline.

mtreinish reviewed (commented):
Overall I'm happy with this direction it makes sense to me, I left a few inline comments from quick things that caught my eye on this first pass. After #8418 merges and this is rebased with tests verifying `transpile()` with control I'll circle back and do a more thorough review.

jakelishman commented on qiskit/transpiler/preset_passmanagers/common.py:
```diff
@@ -51,6 +52,96 @@
+        ("optimization", optimization_method),
+        ("scheduling", scheduling_method),
+    ]:
+        option = stage + "_method"
```
I think I did this just because one of the possible error messages is "the entire {stage} stage is not supported", and that's weird with `scheduling_method`.

jakelishman commented on qiskit/transpiler/preset_passmanagers/common.py:
```diff
@@ -51,6 +52,96 @@
+    "translation_method": ({"translator", "unroller"}, {"synthesis"}),
+    "optimization_method": (set(), set()),
+    "scheduling_method": (set(), {"alap", "asap"}),
+}
```
Done in 5e31bc5d8.

jakelishman commented:
I've added the tests for the new functionality.  They won't pass yet until this PR has #8418 as well, but they should pass immediately after that.

jakelishman commented:
I've added the tests for the new functionality.  They won't pass yet until this PR has #8418 as well, but they should pass immediately after that.

mtreinish commented on releasenotes/notes/transpiler-control-flow-708896bfdb51961d.yaml:
```diff
@@ -0,0 +1,78 @@
+---
+features:
+  - |
+    Control-flow operations are now supported through the transpiler at
+    optimization levels 0 and 1.  One can now construct a circuit such as
```
I might refer to the `transpile()` function here, because the optimization_level argument is on `transpile()`.

jakelishman replied:
Done in a53a4a3fc.

mtreinish commented on releasenotes/notes/transpiler-control-flow-708896bfdb51961d.yaml:
```diff
@@ -0,0 +1,78 @@
+      backend = AerSimulator(method="statevector")
+      transpiled = transpile(qc, backend)
+
+    The available layout methods are "trivial" and "dense".  For routing,
+    "stochastic" and "none" are available.  Translation can be done using
+    "translator" or "unroller".  Optimization levels 2 and 3 are not yet
+    supported with control flow, nor is circuit scheduling, though we intend to
+    expand support for these, and the other layout, routing and translation
+    methods in subsequent releases of Qiskit Terra.
```
I might specify this as `layout_method`, `routing_method`, and ``translation_method` to know these are argument values for transpile (and `generate_preset_pass_manager`).

jakelishman replied:
Done in a53a4a3fc.

mtreinish commented on releasenotes/notes/transpiler-control-flow-708896bfdb51961d.yaml:
```diff
@@ -0,0 +1,78 @@
+    out such transpilations, consider overriding the ``basis_gates`` argument
+    to :func:`.transpile`.
+  - |
+    The following transpiler passes have all been made control-flow aware:
```
While in practice this will always immediately follow the above release note in the list I think it might be good to just add a sentence on what being control-flow aware means in this context.

jakelishman replied:
Done in a53a4a3fc.

mtreinish reviewed (approved):
Overall this LGTM, as you predicted the only comments I had were on the release notes, just a few small ideas and suggestions inline. But not worth blocking over either, we can always work on these in the 0.22.0 final release prep when we edit the release notes too. So don't feel you have to update the release notes now.

jakelishman commented:
Merged with `main` after #8418 merged (and the bug fixes #8843 and #8844), and this is ready for final review and hopefully merge.  I had to correct a couple of tests that I'd borked while writing them without an actual functional implementation, but it should be quite clear from the changeset in 66ed532b7 why it was the tests that were the problem.

mtreinish reviewed (approved):
LGTM, thanks for the updates to the release notes. Looking through the test changes all of those make sense to me. TBH, in my earlier reviews I didn't look too closely at the test code as we were waiting for #8418 still
